src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs

```diff
@@ -196,17 +197,11 @@ public Stream ReadBlob(string dataTableName, string dataColumnName, long rowId)
             // the one the BLOB handle is open on. Calls to sqlite3_blob_read() and 
             // sqlite3_blob_write() for an expired BLOB handle fail with a return code of
             // SQLITE_ABORT.
-            Stream stream = null;
-            RunInTransaction(() =>
+            if (!IsInTransaction)
```

Comment:
this fixing up some existing issue? it used to run RuInTransaction and now it is not? can RunInTransaction see whether it is already running in Transaction and don't run it under transaction?

also. may sql supports nested transaction where inner one is basically NOOP. is that different for sqlite?

Reply:
> this fixing up some existing issue? 

No.  This is necessary for correctness.  See above.  If you want to be able to read/write and you want checksums to match, then you need to be in a transaction so that nothing changes between your operations.

>  can RunInTransaction see whether it is already running in Transaction and don't run it under transaction?

There is no need for this.  We never has nested transactions.  This approach makes sure the caller sets thigns up properly, and it makes sure the caller has to think about it.

Reply:
so why not the top just always open transaction? is there any case we want to do stuff outside of transaction?

Reply:
> is there any case we want to do stuff outside of transaction?

yes.  Most operations can be done outside of a transaction safely.  The only ones i've made sure *must* be done in a transaction are:

1. operations that can lead to corrupt *data* otherwise (i.e. blob reading).
2. operations that would be racey.  i.e. trying to read a stream, but only if the checksums match.  

My goal is to not have anything in this layer do anything that isn't required or appropriate for correctness.  We are attempting to follow the rules and designs of sqlite as closely as possible, that way we're not making assumptions about things work due to stuff we add on after the fact.

Reply:
but my question is why? is there an issue having transaction outside covering everything? otherwise, transaction seems too fragile and easy to get wrong.

Reply:
> but my question is why? 

Because i would really prefer to not have code that does stuff that is not necessary.  *Especially* at this layer.  I've tried to make it very precise and work very correctly with the documented sqlite API.  When we do things taht are unnecessary, it just makes things more confusing IMO because either it will convey the wrong thing to the reader (i.e. that the transaction is necessary), or it will have comments on it saying things like "we don't actually need this, but we're doing it", and eventually we'll be at a point where people will say "i don't want to even touch this because i'm nt sure if it will be a problem"

>  is there an issue having transaction outside covering everything? 

Let me ask you this question:  do you place a `lock (something)` around all assignments in code?  Or do you just use that (and other synchronizations mechanisms) when appropriate because you actually have shared date and/or races to contend with?  

>  transaction seems too fragile and easy to get wrong.

I personally disagree.  I think it's vital that people not just "throw transactions" at the problem, in the same way i dont' think multithreading problems are solved just by adding locks.  Sqlite is *extremely* clear on the guarantees it provides (or not), and i've used transactions *judiciously* in the two places where they are needed for correctness.